runtime/src/alloc.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Minimum buffer between heap and stack
+const MEMORY_GAP: usize = 0x1000;
+
+#[allow(clippy::missing_safety_doc)]
+#[no_mangle]
+pub unsafe extern "C" fn sys_alloc_aligned(bytes: usize, align: usize) -> *mut u8 {
+    #[cfg(target_arch = "riscv32")]
+    extern "C" {
+        // https://lld.llvm.org/ELF/linker_script.html#sections-command
+        static _end: u8;
+    }
+
+    // Pointer to next heap address to use, or 0 if the heap has not yet been
+    // initialized.
+    static mut HEAP_POS: usize = 0;
+
+    // SAFETY: Single threaded, so nothing else can touch this while we're working.
+    let mut heap_pos = HEAP_POS;
+
+    #[cfg(target_arch = "riscv32")]
+    if heap_pos == 0 {
+        heap_pos = &_end as *const u8 as usize;
+    }
+
+    let offset = heap_pos & (align - 1);
+    if offset != 0 {
+        heap_pos = heap_pos
+            .checked_add(align - offset)
+            .expect("No overflow happen");
+    }
+
+    let ptr = heap_pos as *mut u8;
+    heap_pos = heap_pos.checked_add(bytes).expect("No overflow happen");
+
+    // Get the current stack pointer
+    let stack_ptr: usize;
+    unsafe {
+        core::arch::asm!(
+            "mv {}, sp",
+            out(reg) stack_ptr
+        );
+    }
+
+    // Check if the heap is about to clash with the stack
+    if heap_pos + MEMORY_GAP > stack_ptr {
```

Comment:
The addition `heap_pos + MEMORY_GAP` can overflow.

Reply:
Do you think it's necessary, because the start of the heap and stack are in u32, defined in `default.x` and `jolt.x`, 
heap start at `0` or `0x800000`, and stack to `0x40000` or `0x840000` So this addition will overflow unless `heap_pos` is `2**32 - MEMORY_GAP, since the heap range is bounded so no overflow can happen. Similar to the addition with alignment.

Reply:
> Similar to the addition with alignment.
The addition with alignment overflows when `align == 0`.

We have no checks around `sp`. So, `sp` might start from `0x40000`, goes down to `0x32000` ... `0x0004` and underflows to `0xfffffffc`. After that `heap_pos` can increase towards `0xffffff00`. And `heap_pos + MEMORY_GAP` overflows.

(By some luck, I was using Coq to verify an allocator written in C++, and I've thought about these things for weeks.)

Reply:
Ah, I didn't think of the underflow here. Update now.

Reply:
Another way to make this overflow seems to be calling the allocator with `align == 0xffffff00`.

Reply:
Added `checked_add` in d3e02e2

Reply:
Thank you!